DataReader/Model/APIService.swift

```diff
@@ -13,30 +13,44 @@ enum DataManagerError: Error {
     case invalidJson
 }
 
-protocol FetchManagerDelegate {
-    func requestCompleted(data: [String]?, error: DataManagerError?)
+typealias DataTaskResult = (Data?, URLResponse?, Error?) -> Void
+
+protocol URLSessionDataTaskProtocol {
+    func resume()
 }
 
-class APIService {
-    static func requestData(delegate: FetchManagerDelegate)  {
-        
-        let jsonUrlString = "https://private-5e934f-datatextapi.apiary-mock.com/data"
-        guard let url = URL(string: jsonUrlString) else { return }
+protocol URLSessionProtocol {
+    typealias DataTaskResult = (Data?, URLResponse?, Error?) -> Void
+    
+    func dataTask(with url: URL,
+                  completionHandler: @escaping DataTaskResult) -> URLSessionDataTaskProtocol
+}
+
+extension URLSessionDataTask: URLSessionDataTaskProtocol { }
+
+extension URLSession: URLSessionProtocol {
+    func dataTask(with url: URL,
+                  completionHandler: @escaping DataTaskResult) -> URLSessionDataTaskProtocol {
         
-        URLSession.shared.dataTask(with: url) { (data, response, err) in
-            // TODO: check err and response status
-            guard let data = data else { return }
-            
-            do {
-                let json = try JSONSerialization.jsonObject(with: data) as? [String]
-                if let stringArray = json {
-                    delegate.requestCompleted(data: stringArray, error: nil)
-                }
-            } catch {
-                delegate.requestCompleted(data: nil, error: .invalidJson)
-            }
-            
-            }.resume()
+        return (dataTask(with: url, completionHandler: completionHandler) as URLSessionDataTask) as URLSessionDataTaskProtocol
     }
+}
 
+class APIService: APIServiceProtocol {
+    
+    private let session: URLSessionProtocol!
+    
+    init(session: URLSessionProtocol = URLSession.shared ) {
+        self.session = session
+    }
+    
+    func requestData(url: URL, callback: @escaping completeClosure)  {
+        self.session.dataTask(with: url) { (data, response, error) in
+            if let _ = error {
+                callback(nil, DataManagerError.unknown)
+            } else {
+                callback(data, nil)
+            }
+        }.resume()
```

Comment:
if you pass your own session instead shared then you will have memory leak
details: https://stackoverflow.com/questions/39409357/nsurlsession-http-2-memory-leak/46613760#46613760